content/docs/rfcs/44/README.md

```diff
@@ -0,0 +1,261 @@
+It attaches the `settlement_response` to its `HistoryRPC` and sends it back to the querying node. 
+- The queried node keeps listening to the Waku service contract to get an event about the payment of the `request_id`. 
+Once the payment is received, it proceeds with the normal Store protocol flow with a slight difference. 
+That is, it attaches `settlement_response` to its `HistoryRPC` with all the data fields as specified in the step above. 
+It also increases the `history_retrieved` of the `settlement_response` with the size of the  `messages` in the `HistoryResponse`.
+This is to indicate how many messages are transferred so far and how much is left.
+
+
+# Service Credential Structure 
+<!-- Each service token is associated with a specific value namely `amount` determined by its owner. 
+The `amount` is determined based on the cost of the history query.  -->
+Each service credential is structured as below:
+``` 
+commitment = poseidonHash(amount,  pubkey = poseidonHash(privkey),  blinding)
+```
+`amount` determines the value of the credential in terms of Ether.
+The `privkey` is a random `32` bytes value. 
+The `blinding` is a random `32` bytes blinding factor.
+
+There is also an encryption key `enckey` derived from the `privkey`. 
+<!-- the encryption key derivation to be specified -->
+
+## Service Credential Nullifier 
+
+Each service credential is associated with a nullifier that is computed as follows:
+
+```
+nullifier = poseidonHash(commitment, index || 0, signature)
+```
+where `index` indicates the position of the `commitment` in the Merkle tree,
+and the `signature` is generated over the `commitment` and the `index` as:
+```
+signature = poseidonHash(privkey, commitment, index)
+```
+
+
+# Waku Service Contract
+
+## Registered store nodes
+Nodes offering store protocol who are willing to use the private settlement MUST be registered in the Waku service contract.
+The list of the Ethereum address of the registered nodes is available in the state of the contract.
+
+## Merkle tree
+
+The set of registered service credentials is  represented by a Merkle tree  where the service credential commitments constitute the leaves. 
+The Merkle tree has an append only structure i.e., leaves are only added but not deleted.
+The list of the last 20 valid MT roots are also available in the state of the contract.
+
+## Spent Credentials
+The contract keeps track of the spent credentials by maintaining their nullifiers in the state of the contract.
+Credentials whose nullifiers belong to this list will not be redeemable.
+
+## Transaction
+Users interact with the contract to either `purchase` a credential or `transfer` a portion of their credential for some storage service. 
+Regardless of the type of the transaction i.e., being a `purchase` or `transfer`, all the transactions follow a general structure which is inspired by bitcoin UTXO idea. 
+
+Each service credential is treated as a UTXO.
```

Comment:
This sounds interesting, but I'm not quite sure I follow. Perhaps motivation and illustration would be useful?

(Or maybe I just need to read it more carefully 😅 )

Reply:
Is this idea from somewhere in particular? Is this how Nova works under the hood somehow?

Reply:
Thanks for the comments @oskarth 
This write up was not ready for review :) I will soon finish up my revisions and make it ready!
The quick response is that 
-> UTXO is how Nova works under the hood
-> Registration is only needed for the store nodes, not the querying node, and the reason is to later be able to add reputation, also to make it hard for the users of the system to use the contract for arbitrary payments (outside of the store protocol)

I am adding more details on the rationale of design choices, hopefully, they would address your concerns